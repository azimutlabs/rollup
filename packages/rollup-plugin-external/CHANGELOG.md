# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.1.0](https://github.com/azimutlabs/rollup/compare/@azimutlabs/rollup-plugin-external@0.1.0-alpha.0...@azimutlabs/rollup-plugin-external@0.1.0) (2021-02-26)

**Note:** Version bump only for package @azimutlabs/rollup-plugin-external





# 0.1.0-alpha.0 (2021-02-24)


### Bug Fixes

* **rollup-plugin-external:** change nullish operators ([987fae2](https://github.com/azimutlabs/rollup/commit/987fae25b023ae00dc9144f26efec0a392c52896))


### Features

* add rollup plugin external ([bbc8fce](https://github.com/azimutlabs/rollup/commit/bbc8fce7f64bb99c12a57b8834821c5bf04b5a2d))
