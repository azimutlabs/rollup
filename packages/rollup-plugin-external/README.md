<h1 align="center">
  <a target="_blank" href="https://alabs.team">
    üç£
    <img
      height="22.5"
      src="https://raw.githubusercontent.com/azimutlabs/logos/master/little_logo.png"
      alt="azimutlabs logo"
    />
    /rollup-plugin-external
  </a>
</h1>

<p align="center">Rollup plugin to exclude external dependencies from bundle</p>

<p align="center">
  <a href="https://github.com/azimutlabs/rollup/actions?query=workflow%3A%22Lint+and+Test%22">
    <img
      src="https://github.com/azimutlabs/rollup/workflows/Lint%20and%20Test/badge.svg"
      alt="azimutlabs/rollup repository github workflow status"
    />
  </a>
  <a href="https://github.com/azimutlabs/rollup/blob/master/LICENSE">
    <img
      src="https://img.shields.io/github/license/azimutlabs/rollup?label=License"
      alt="azimutlabs/rollup repository license"
    />
  </a>
  <a href="https://www.npmjs.com/package/@azimutlabs/rollup-plugin-external">
    <img
      src="https://img.shields.io/npm/v/@azimutlabs/rollup-plugin-external?color=blue&logo=npm&label="
      alt="@azimutlabs/rollup-plugin-external"
    />
  </a>
</p>

## Installation
```shell
$ yarn add --save-dev @azimutlabs/rollup-plugin-external
```

## Usage
```javascript
// rollup.config.js
import external from '@azimutlabs/rollup-plugin-external';

export default {
  input: 'index.js',
  plugins: [
    external({
      // Default options.
      useDependencies: true,
      usePeerDependencies: true,
      useBuiltins: true,
      checkForBabelMacro: false,
      packagePath: process.cwd(),
    }),
  ],
};
```

Plugin does not override rollup's [external](https://rollupjs.org/guide/en/#external) option.
Instead, it firstly looks for external, then applies plugin's core functionality
```javascript
// rollup.package.js
import external from '@azimutlabs/rollup-plugin-external';

export default {
  input: 'index.js',
  plugins: [external()],
  external: id => id.includes('fp-ts'),
};
```

## Options
| Name | Description | Type | Default |
|------|-------------|------|---------|
|`useDependencies`|Indicates whether to include `dependencies` from `package.json`|`boolean`|`true`|
|`usePeerDependencies`|Indicates whether to include `peer dependencies` from `package.json`|`boolean`|`true`|
|`useBuiltins`|Indicates whether to include **Node.js** builtin modules|`boolean`|`true`|
|`checkForBabelMacro`|Indicates whether to check `package.json` for [**Babel Macro**](https://github.com/kentcdodds/babel-plugin-macros) packages|`boolean`|`false`|
|`packagePath`|Directory or path to `package.json`. If not provided, looks for closest `package.json` starting from current directory|`string`|`process.cwd()`|

## Acknowledgement
This plugin was inspired by [`rollup-plugin-auto-external`](https://github.com/stevenbenisek/rollup-plugin-auto-external)

## Contributing
Any PR is welcomed by our **@js-opensource** team.
Check out our [contributing](../../CONTRIBUTING.md) guidelines for more info.

## License
[![azimutlabs rollup config license](https://img.shields.io/github/license/azimutlabs/rollup?label=as%20always&color=informational)](../../LICENSE)
