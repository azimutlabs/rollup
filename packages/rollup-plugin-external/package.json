{
  "name": "@azimutlabs/rollup-plugin-external",
  "version": "0.0.0",
  "description": "Rollup plugin to exclude external dependencies from bundle",
  "keywords": [
    "plugin",
    "rollup",
    "rollup-plugin",
    "alabs",
    "azimutlabs"
  ],
  "homepage": "https://github.com/azimutlabs/rollup#readme",
  "bugs": {
    "url": "https://github.com/azimutlabs/rollup/issues",
    "email": "npm@alabs.team"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/azimutlabs/rollup.git",
    "directory": "packages/rollup-plugin-external"
  },
  "license": "MIT",
  "author": "Azimut Labs <opensource@alabs.team>",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "compile": "yarn tsc"
  },
  "dependencies": {
    "builtin-modules": "^3.2.0",
    "is-valid-path": "^0.1.1",
    "lodash.concat": "^4.5.0",
    "read-pkg-up": "^7.0.1"
  },
  "devDependencies": {
    "@types/builtin-modules": "^3.1.1",
    "@types/is-valid-path": "^0.1.0",
    "@types/lodash.concat": "^4.5.6",
    "@types/read-pkg-up": "^6.0.0",
    "rollup": "^2.38.0",
    "typescript": "^4.1.4"
  },
  "publishConfig": {
    "access": "public"
  }
}
