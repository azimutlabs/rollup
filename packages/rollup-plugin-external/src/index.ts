export { rollupPluginExternal as default, rollupPluginExternal } from './plugin';
export * from './types/RollupPluginExternalOptions';
