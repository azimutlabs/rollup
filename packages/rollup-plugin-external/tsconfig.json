{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": "./src",
    "outDir": "./lib",
    "target": "es5",
    "module": "commonjs"
  },
  "include": [
    "src"
  ]
}
