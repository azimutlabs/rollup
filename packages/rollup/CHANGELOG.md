# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0-alpha.0 (2021-02-25)


### Features

* **rollup:** add useWorkspaces function ([#19](https://github.com/azimutlabs/rollup/issues/19)) ([324c47f](https://github.com/azimutlabs/rollup/commit/324c47f1b60fe1a489950edbe1e39ff200af198f))
* add rollup collect ([#8](https://github.com/azimutlabs/rollup/issues/8)) ([f43084a](https://github.com/azimutlabs/rollup/commit/f43084a1501c3faeb12f155272a79fee835bb67a))
