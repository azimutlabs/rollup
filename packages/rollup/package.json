{
  "name": "@azimutlabs/rollup",
  "version": "0.1.4",
  "description": "Collect Rollup configurations into a singular array",
  "keywords": [
    "config",
    "rollup",
    "alabs",
    "azimutlabs"
  ],
  "homepage": "https://github.com/azimutlabs/rollup#readme",
  "bugs": {
    "url": "https://github.com/azimutlabs/rollup/issues",
    "email": "npm@alabs.team"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/azimutlabs/rollup.git",
    "directory": "packages/rollup"
  },
  "license": "MIT",
  "author": "Azimut Labs <opensource@alabs.team>",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "compile": "yarn tsc"
  },
  "dependencies": {
    "glob": "^7.1.6",
    "read-pkg": "^5.2.0",
    "read-pkg-up": "^7.0.1"
  },
  "peerDependencies": {
    "rollup": "^2.15"
  },
  "devDependencies": {
    "@types/glob": "^7.1.3",
    "@types/node": "^14.14.31",
    "@types/read-pkg": "^5.1.0",
    "@types/read-pkg-up": "^6.0.0",
    "rollup": "^2.38.0",
    "typescript": "^4.1.4"
  },
  "publishConfig": {
    "access": "public"
  }
}
