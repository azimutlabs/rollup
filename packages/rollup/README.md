<h1 align="center">
  <a target="_blank" href="https://alabs.team">
    üç£
    <img
      height="22.5"
      src="https://raw.githubusercontent.com/azimutlabs/logos/master/little_logo.png"
      alt="azimutlabs logo"
    />
    /rollup
  </a>
</h1>

<p align="center">Collect Rollup configurations into a singular array</p>

<p align="center">
  <a href="https://github.com/azimutlabs/rollup/actions?query=workflow%3A%22Lint+and+Test%22">
    <img
      src="https://github.com/azimutlabs/rollup/workflows/Lint%20and%20Test/badge.svg"
      alt="azimutlabs/rollup repository github workflow status"
    />
  </a>
  <a href="https://github.com/azimutlabs/rollup/blob/master/LICENSE">
    <img
      src="https://img.shields.io/github/license/azimutlabs/rollup?label=License"
      alt="azimutlabs/rollup repository license"
    />
  </a>
   <a href="https://www.npmjs.com/package/@azimutlabs/rollup">
     <img
       src="https://img.shields.io/npm/v/@azimutlabs/rollup?color=blue&logo=npm&label="
       alt="@azimutlabs/rollup"
     />
   </a>
</p>

## Installation
 ```shell
 $ yarn add -D @azimutlabs/rollup
 ```

## Known issues
+ `collect` does not work with `esm`, so in order to use it, it is required to use `CommonJS` modules
  in rollup config. See [#41](https://github.com/azimutlabs/rollup/issues/41) for more info.

## Usage

### `collect`
**Collect configurations by the given array of glob patterns.**
```typescript
/**
 * @param patterns - packages glob patterns.
 * @param dirname - working directory root. defaults to process.cwd()
 */
function collect(patterns: readonly string[], dirname?: string): Promise<readonly RollupOptions[]>;
```

Given example:
```javascript
// packages/ui/rollup.config.js
import compose from '@azimutlabs/rollup-config';
import babel from '@azimutlabs/rollup-config-babel';

export default compose(__dirname, babel());
```
```javascript
// rollup.config.js
import { collect } from '@azimutlabs/rollup';

export default collect(['packages/*/rollup.config.js']);
```
...will result in:
```javascript
// rollup.config.js
// packages/ui/lib/
//   index.es.js - output from babel
export default [
  { /* packages/ui/rollup.config.js */ }
];
```

### `fromWorkspaces`
**Get packages glob patterns from the `workspaces` field in the nearest `package.json`**
```typescript
/**
 * @param pattern - rollup config files pattern. defaults to 'rollup.config.js'
 */
function fromWorkspaces(pattern?: string): readonly string[];
```

This is recommended usage when using yarn/npm
[workspaces](https://classic.yarnpkg.com/en/docs/workspaces/) feature.
```json5
// package.json
{
  "workspaces": [
    "packages/*"
  ]
}
```
```javascript
// rollup.config.js
import { collect, fromWorkspaces } from '@azimutlabs/rollup';

export default collect(
  // Final collect scope will be: ['packages/*/rollup.config.js']
  fromWorkspaces(/* glob pattern for config files, e.g. rollup.config*.js */)
);
```

## Contributing
Any PR is welcomed by our **@js-opensource** team.
Check out our [contributing](../../CONTRIBUTING.md) guidelines for more info.

## License
[![azimutlabs rollup config license](https://img.shields.io/github/license/azimutlabs/rollup?label=as%20always&color=informational)](../../LICENSE)
