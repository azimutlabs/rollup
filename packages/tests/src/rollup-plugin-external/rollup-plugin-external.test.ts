import type { RollupPluginExternalOptions } from '@azimutlabs/rollup-plugin-external';
import { rollupPluginExternal } from '@azimutlabs/rollup-plugin-external';
import commonjs from '@rollup/plugin-commonjs';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import typescript from '@rollup/plugin-typescript';
import path from 'path';
import type { InputOptions, Plugin } from 'rollup';
import { rollup } from 'rollup';
import nodePolyfills from 'rollup-plugin-node-polyfills';

const mockRollup = async (
  rollupOptions: InputOptions,
  pluginOptions?: RollupPluginExternalOptions
): Promise<void> => {
  const bundle = await rollup({
    ...rollupOptions,
    plugins: [
      typescript({ tsconfig: false, allowSyntheticDefaultImports: true }),
      rollupPluginExternal(pluginOptions),
      nodePolyfills() as Plugin,
      nodeResolve(),
      commonjs(),
    ],
  });
  const result = await bundle.generate({ format: 'esm' });
  const [{ code }] = result.output;
  expect(code).toMatchSnapshot();
};

describe('plugin-external', () => {
  it('should have a name', () => {
    expect(rollupPluginExternal().name).toEqual('@azimutlabs/rollup-plugin-external');
  });

  it('should handle dependencies', async () => {
    await mockRollup({ input: path.resolve(__dirname, './example.ts') });
  });

  it('should handle disabling dependencies', async () => {
    await mockRollup(
      { input: path.resolve(__dirname, './example.ts') },
      { useDependencies: false }
    );
  });

  it('should handle disabling builtins', async () => {
    await mockRollup({ input: path.resolve(__dirname, './example.ts') }, { useBuiltins: false });
  });

  it('should handle disabling peer dependencies', async () => {
    await mockRollup(
      { input: path.resolve(__dirname, './example.ts') },
      { usePeerDependencies: false }
    );
  });
});
