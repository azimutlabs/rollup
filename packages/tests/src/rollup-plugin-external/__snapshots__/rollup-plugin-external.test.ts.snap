// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-external should handle dependencies 1`] = `
"import { right, reduce } from 'fp-ts/Either';
import { pipe } from 'fp-ts/function';
import isValid from 'is-valid-path';
import { cwd } from 'process';

var example = function () {
    var startWith = 'prefix';
    var dir = cwd();
    var concat = function (aa, bb) { return dir + \\":\\" + aa + \\":\\" + bb; };
    return isValid(pipe(right('e'), reduce(startWith, concat)));
};

export { example };
"
`;

exports[`plugin-external should handle disabling builtins 1`] = `
"import { right, reduce } from 'fp-ts/Either';
import { pipe } from 'fp-ts/function';
import isValid from 'is-valid-path';

// shim for using process in browser

function cwd () { return '/' }

var example = function () {
    var startWith = 'prefix';
    var dir = cwd();
    var concat = function (aa, bb) { return dir + \\":\\" + aa + \\":\\" + bb; };
    return isValid(pipe(right('e'), reduce(startWith, concat)));
};

export { example };
"
`;

exports[`plugin-external should handle disabling dependencies 1`] = `
"import isValid from 'is-valid-path';
import { cwd } from 'process';

/**
 * @since 2.0.0
 */
function pipe(a, ab, bc, cd, de, ef, fg, gh, hi, ij, jk, kl, lm, mn, no, op, pq, qr, rs, st) {
    switch (arguments.length) {
        case 1:
            return a;
        case 2:
            return ab(a);
        case 3:
            return bc(ab(a));
        case 4:
            return cd(bc(ab(a)));
        case 5:
            return de(cd(bc(ab(a))));
        case 6:
            return ef(de(cd(bc(ab(a)))));
        case 7:
            return fg(ef(de(cd(bc(ab(a))))));
        case 8:
            return gh(fg(ef(de(cd(bc(ab(a)))))));
        case 9:
            return hi(gh(fg(ef(de(cd(bc(ab(a))))))));
        case 10:
            return ij(hi(gh(fg(ef(de(cd(bc(ab(a)))))))));
        case 11:
            return jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a))))))))));
        case 12:
            return kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a)))))))))));
        case 13:
            return lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a))))))))))));
        case 14:
            return mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a)))))))))))));
        case 15:
            return no(mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a))))))))))))));
        case 16:
            return op(no(mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a)))))))))))))));
        case 17:
            return pq(op(no(mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a))))))))))))))));
        case 18:
            return qr(pq(op(no(mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a)))))))))))))))));
        case 19:
            return rs(qr(pq(op(no(mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a))))))))))))))))));
        case 20:
            return st(rs(qr(pq(op(no(mn(lm(kl(jk(ij(hi(gh(fg(ef(de(cd(bc(ab(a)))))))))))))))))));
    }
    return;
}

// -------------------------------------------------------------------------------------
// guards
// -------------------------------------------------------------------------------------
/**
 * Returns \`true\` if the either is an instance of \`Left\`, \`false\` otherwise.
 *
 * @category guards
 * @since 2.0.0
 */
var isLeft = function (ma) { return ma._tag === 'Left'; };
/**
 * Constructs a new \`Either\` holding a \`Right\` value. This usually represents a successful value due to the right bias
 * of this structure.
 *
 * @category constructors
 * @since 2.0.0
 */
var right = function (a) { return ({ _tag: 'Right', right: a }); };
/**
 * Left-associative fold of a structure.
 *
 * @example
 * import { pipe } from 'fp-ts/function'
 * import * as E from 'fp-ts/Either'
 *
 * const startWith = 'prefix'
 * const concat = (a: string, b: string) => \`\${a}:\${b}\`
 *
 * assert.deepStrictEqual(
 *   pipe(E.right('a'), E.reduce(startWith, concat)),
 *   'prefix:a',
 * )
 *
 * assert.deepStrictEqual(
 *   pipe(E.left('e'), E.reduce(startWith, concat)),
 *   'prefix',
 * )
 *
 * @category Foldable
 * @since 2.0.0
 */
var reduce = function (b, f) { return function (fa) {
    return isLeft(fa) ? b : f(b, fa.right);
}; };

var example = function () {
    var startWith = 'prefix';
    var dir = cwd();
    var concat = function (aa, bb) { return dir + \\":\\" + aa + \\":\\" + bb; };
    return isValid(pipe(right('e'), reduce(startWith, concat)));
};

export { example };
"
`;

exports[`plugin-external should handle disabling peer dependencies 1`] = `
"import { right, reduce } from 'fp-ts/Either';
import { pipe } from 'fp-ts/function';
import { cwd } from 'process';

/*!
 * is-extglob <https://github.com/jonschlinkert/is-extglob>
 *
 * Copyright (c) 2014-2015, Jon Schlinkert.
 * Licensed under the MIT License.
 */
var isExtglob = function isExtglob(str) {
  return typeof str === 'string'
    && /[@?!+*]\\\\(/.test(str);
};

/*!
 * is-glob <https://github.com/jonschlinkert/is-glob>
 *
 * Copyright (c) 2014-2015, Jon Schlinkert.
 * Licensed under the MIT License.
 */

var isGlob = function isGlob(str) {
  return typeof str === 'string'
    && (/[*!?{}(|)[\\\\]]/.test(str)
     || isExtglob(str));
};

/*!
 * is-invalid-path <https://github.com/jonschlinkert/is-invalid-path>
 *
 * Copyright (c) 2015, Jon Schlinkert.
 * Licensed under the MIT License.
 */


var re = /[‘“!#$%&+^<=>\`]/;

var isInvalidPath = function (str) {
  return (typeof str !== 'string') || isGlob(str) || re.test(str);
};

/*!
 * is-valid-path <https://github.com/jonschlinkert/is-valid-path>
 *
 * Copyright (c) 2015 Jon Schlinkert, contributors.
 * Licensed under the MIT license.
 */



var isValidPath = function (str) {
  return isInvalidPath(str) === false;
};

var example = function () {
    var startWith = 'prefix';
    var dir = cwd();
    var concat = function (aa, bb) { return dir + \\":\\" + aa + \\":\\" + bb; };
    return isValidPath(pipe(right('e'), reduce(startWith, concat)));
};

export { example };
"
`;
